web/src/dialog_widget.ts

```diff
@@ -42,15 +41,55 @@ import * as ui_report from "./ui_report";
  *          to DOM, it can do so by passing a post_render hook.
  */
 
-export function hide_dialog_spinner() {
+type WidgetConfig = {
+    html_heading: string;
+    html_body: string;
+    on_click: (e: unknown) => void;
+    html_submit_button?: string;
+    close_on_submit?: boolean;
+    focus_submit_on_open?: boolean;
+    help_link?: string;
+    id?: string;
+    single_footer_button?: boolean;
+    form_id?: string;
+    validate_input?: (e: unknown) => boolean;
+    on_show?: () => void;
+    on_shown?: () => void;
+    on_hide?: () => void;
+    on_hidden?: () => void;
+    post_render?: () => void;
+    loading_spinner?: boolean;
+};
+
+// TODO: This type should probably be exported from channel.ts once
+// that's converted to TypeScript.
+type AjaxRequest = ({
+    url,
+    data = {},
+    success,
+    error,
+}: {
+    url: string;
+    data?: Record<string, never>;
+    success(response_data?: string): void;
+    error(xhr?: JQuery.jqXHR): void;
+}) => void;
+
+type RequestOpts = {
+    failure_msg_html?: string;
+    success_continuation?: (response_data?: string) => void;
+    error_continuation?: (xhr?: JQuery.jqXHR) => void;
```

Comment:
`response_data` and `xhr` should not be marked optional here either. `response_data` is not a `string`.